clis/teliod/src/daemon.rs

```diff
@@ -0,0 +1,66 @@
+//! Code for general daemon handling utilities and abstracting away related dependencies.
+
+use std::path::PathBuf;
+
+use anyhow::Result;
+use sysinfo::{ProcessRefreshKind, RefreshKind, System};
+
+/// Struct for packaging together the functions for managing the daemon.
+/// Currently does not need any members, since most of the data is stored
+/// in files anyway.
+pub trait Daemon {
+    /// Daemon name, used for naming daemon files
+    const NAME: &'static str;
+
+    /// Helper function for getting the path to the daemon's working directory.
+    /// The directory is used for storing *.pid and *.socket files.
+    ///
+    /// # Returns
+    ///
+    /// Result containing the path of the daemons work directory.
+    fn get_wd_path() -> Result<PathBuf>;
+
+    /// Checks whether the daemon is currently running.
+    /// Won't work if pid file had been tampered with or it's location changed.
+    ///
+    /// # Returns
+    ///
+    /// Result containing a bool. True means that the daemon is running.
+    fn is_running() -> Result<bool> {
+        let system = System::new_with_specifics(
+            RefreshKind::new().with_processes(ProcessRefreshKind::everything()),
+        );
+
+        match std::fs::read_to_string(Self::get_pid_path()?) {
```

Comment:
I guess this part is inherently racy, e.g. if `teliod` crashes or is killed abruptly, pid file will be left there, and then there is always a chance that some other process will take previous pid. While I agree that this part is quite an edge-case, but maybe there is some usefulness to rely on unix socket for that. E.g. if new daemon process is being started and unix socket exists -> issue an error. In case of a client process - failure to open unix socket could indicate that daemon is not running.

This of course means, that we do need to handle unix socket cleanup.